src/com/ikhokha/techcheck/Main.java

```diff
@@ -4,42 +4,73 @@
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
 
 public class Main {
 
-	public static void main(String[] args) {
-		
-		Map<String, Integer> totalResults = new HashMap<>();
-				
-		File docPath = new File("docs");
-		File[] commentFiles = docPath.listFiles((d, n) -> n.endsWith(".txt"));
-		ArrayList<CommentMetricProcessor> commentMetrics= new ArrayList<>(Arrays.asList(new SearchLinkUrl("SPAM"),new SearchOccurance( "QUESTIONS", "?"),new SearchOccurance( "SHAKER_MENTIONS", "Shaker"),new SearchOccurance( "MOVER_MENTIONS", "Mover"),new ShortComments( "SHORTER_THAN_15")));
-		
-		for (File commentFile : commentFiles) {
-			
-			CommentAnalyzer commentAnalyzer = new CommentAnalyzer(commentFile,commentMetrics);
-			Map<String, Integer> fileResults = commentAnalyzer.analyze();
-			addReportResults(fileResults, totalResults);
-						
+	public static void main(String[] args) throws InterruptedException, ExecutionException {
+		final int MaxNumberOfThreads = Runtime.getRuntime().availableProcessors() / 2;
+		Map<String, Integer> totalProcessedOutputMetricsResults = new HashMap<>();
```

Comment:
Multithreading has no noticeable effect when file number is low and the content is very little. After going past 1000 files half of the number of runtime available processors to outperform most most setups. The system design is chosen after some testing and thorough research .